T1505.003 Server Software Component: Web Shells (#23)
tired-labs/techniques: reports/trr0000/win/README.md

# File-Based Web Shell Execution via IIS

## Metadata

| Key          | Value                                          |
|--------------|------------------------------------------------|
| ID           | TRR0000                                        |
| External IDs | [T1505.003]                                    |
| Tactics      | Persistence                                    |
| Platforms    | Windows                                        |
| Contributors | John McGuinness                                |

### Scope Statement

This TRR covers file-based web shells executed through IIS
handler mappings on Windows.

| Excluded Item | Rationale |
|---|---|
| Fileless or memory-only web shells | Different essential operations |
| Other web servers on Windows (Apache, nginx, Tomcat) | Different essential operations |
| ASP.NET Core on IIS | Different execution model (IIS proxies to Kestrel) |

> | PHP on IIS via FastCGI | Different execution model (via php-cgi.exe) |

Reviewer: A table lends itself to adding line after line easily. We don't want to fall into the slippery slope of defining too many out-of-scope items (ie Linux, Apache, Commodore 64, Game Boy Color, Samsung refrigerators).

Lets keep this section to discussing what IS in the document and for important omissions (like `ASP.NET Core`) we can add it into the paragraph.

Author: Understood! I was back and forth on the table, on one hand I felt it was easier to read but I see what you're saying. I also see why I can take off the Fileless/memory and non-windows stuff from the exclusions. It's kind of  obvious being how different they are. I'll make a correction and try again.

## Technique Overview

A web shell is a malicious script placed on a web server that gives an attacker
remote command execution through the server's HTTP request-handling mechanisms.
File-based web shells persist as script files on disk and execute through the
server's handler mappings. They are attractive to adversaries because they
survive reboots, blend with normal web traffic on ports 80/443, and require no

> additional software.

Reviewer 2: ```suggestion
additional dependencies.
```

Apologies if this is too nitpicky, but I would argue the webshell itself is additional software, as opposed to something like LOLBAS where the binary lives on disk already.

## Technical Background

### IIS Architecture

Internet Information Services (IIS) is Microsoft's web server, included with

> Windows Server. Every web shell on IIS must pass through the same request

Reviewer 3: Very nit-picky comment here, but maybe it's worth clarifying that it's included, but not enabled, by default in Win Server?

Author: Nit-picky is what refines these things, right? That's a fair point, thanks!

processing pipeline regardless of the shell's implementation.

#### HTTP.sys

All HTTP/HTTPS traffic destined for IIS passes through HTTP.sys, a kernel-mode
driver that listens on configured ports (typically 80 and 443). HTTP.sys
handles connection management, caching, and request routing. When a request
requires application code, HTTP.sys routes it to the appropriate application
pool based on site bindings. The Windows Process Activation Service
(WAS) mediates between HTTP.sys and the worker process, managing application
pool lifecycle and process activation. HTTP.sys and WAS operate at the kernel
and system service level respectively, below the reach of most telemetry

> sources.

Reviewer: ```suggestion
All HTTP/HTTPS traffic destined for IIS passes through `HTTP.sys`, a kernel-mode
driver that listens on configured ports (typically 80 and 443). `HTTP.sys`
handles connection management, caching, and request routing. When a request
requires application code, `HTTP.sys` routes it to the appropriate application
pool based on site bindings. The Windows Process Activation Service
(WAS) mediates between `HTTP.sys` and the worker process, managing application
pool lifecycle and process activation. `HTTP.sys` and WAS operate at the kernel
and system service level respectively, below the reach of most telemetry
sources.
```

Adding backticks around filenames for readability

#### Application Pools

An application pool is an isolation boundary within IIS. Each pool runs as its
own instance of the worker process `w3wp.exe`, with its own security identity
and resource limits. The pool's configured identity determines the security

> context under which all code executes — including web shell code.

Reviewer 2: ```suggestion
context under which all code executes - including web shell code.
```

Em dashes are unicode and I don't like unicode in markdown files. Others may disagree tho.

By default, pools run under low-privilege virtual accounts (e.g.,

> `IIS AppPool\DefaultAppPool`). Administrators sometimes configure pools to use

Reviewer: ```suggestion
`IIS AppPool\DefaultAppPool`). Administrators are able to configure pools to use
```

Removing the ambiguous `sometimes` verbiage

higher-privilege accounts such as `NetworkService`, domain service accounts, or
`LocalSystem`. The web shell inherits whatever permissions the pool identity
has.

#### The IIS Worker Process (w3wp.exe)

The worker process `w3wp.exe` is the most important component for understanding
web shell execution. It loads and executes requested files using the
appropriate scripting engine, runs under the application pool's identity, and
can spawn child processes.

When a web shell executes OS commands by starting external programs (such as
`cmd.exe` or `powershell.exe`), those programs appear as child processes of
`w3wp.exe`. Under normal operations, `w3wp.exe` rarely spawns command

> interpreters.

Reviewer: Is there a source for this claim about "rarely" spawning command interpreters? May not be needed but just curious

Author: Yeah that's basically hearsay. I probably read that it doesn't normally happen but that could not be true for any given environment. I'll remove it.

Reviewer 3: Well maybe it doesn't need to be removed? I'll defer to the other people here who are smarter than I am, but it might be a detection consideration? I could see it varying in different environments, but it might be worth investigating in your own env. how often IIS web apps are spawning command interpreters, etc.

However, web shells that operate exclusively through .NET framework APIs (e.g.,
`System.IO.File`, `System.Net.WebClient`, `System.Data.SqlClient`) execute
entirely within `w3wp.exe` without spawning a child process. This distinction
is the basis for two of the procedures in this TRR.

### Handler Mappings

IIS uses handler mappings to determine how to process a requested file. When a
request arrives, IIS examines the file extension to determine whether the file
should be served as static content or processed by a scripting engine.

On a default IIS installation with ASP.NET enabled, the following extensions
are mapped to executable handlers:

- `.aspx` — ASP.NET engine
- `.asp` — Classic ASP engine
- `.ashx` — ASP.NET generic handler

> - `.asmx` — ASP.NET web service handler

Reviewer: You list `.asp`, `.aspx`, `.ashx`, and `.asmx` under "a default IIS installation with ASP.NET enabled" but Classic ASP (which handles `.asp`) and ASP.NET are separate IIS role services. 

Installing the ASP.NET role does not install Classic ASP, and vice versa. On a default install with only ASP.NET enabled, `.asp` would not have a functioning handler mapping.

Source: https://learn.microsoft.com/en-us/iis/application-frameworks/running-classic-asp-applications-on-iis-7-and-iis-8/classic-asp-not-installed-by-default-on-iis

Author: This is one of those I can slap myself 3 weeks ago. I read it, forgot I read it, never made the connection to fix it. I did know that Classic ASP is not installed by default, anymore. I can rewrite this section so it's more accurate.

> - `.asmx` — ASP.NET web service handler

Reviewer 2: ```suggestion
- `.aspx` - ASP.NET engine
- `.asp` - Classic ASP engine
- `.ashx` - ASP.NET generic handler
- `.asmx` - ASP.NET web service handler
```

More em dashes

IIS also supports Server-Side Includes (SSI) via `ssinc.dll`. If enabled,
files with extensions `.shtml`, `.stm`, and `.shtm` can execute directives
such as `<!--#exec cmd="command">`. The `#exec cmd=` directive spawns a child
process and follows the same essential operations as Procedure A. Other SSI
directives (`#include`, `#echo`) are handled internally by `ssinc.dll` without
passing through a scripting engine; while these are in-process operations, they

> do not map to Procedure B.

Reviewer: You claim:
>  while these are in-process operations, they do not map to Procedure B.

If procedure B is `Web Shell with In-Process Execution`. can you elaborate why it does not map?

Author: Oh my lordy. I've re-read this specific thing like 6 times. I'm still not fully sure understand it. I THINK it's because #include and #echo can't execute code, they only do a specific thing and can't perform calls, executions, or queries outside of read or output a server variable for #include and #echo respectively, also there is no scripting engine running these. So basically it never hits the "execute code" part of the procedures because there's no code to execute. If it mapped to Proc B, the attacker code would be executed by a scripting engine.

SO I think maybe I just wasn't clear enough. I'll rewrite this and see what you think then.

https://learn.microsoft.com/en-us/iis/configuration/system.webserver/serversideinclude

For a web shell to function under default handler mappings, it must use a file
extension already mapped to an executable handler. This is an essential and

> immutable constraint — unless the attacker modifies the handler mappings

Reviewer 2: ```suggestion
immutable constraint - unless the attacker modifies the handler mappings
```

Another em dash.

> themselves (see web.config section below).

Reviewer: This paragraph is wonky. First we talk about default handler mappings when the previous section discussed enabling Server-Side Includes, which is not a default setting it sounds like.

Then you mention an "immutable contraint" followed by a way that it is mutable.

Finally, you say "unless the attacker modifies", but theoretically the administrator can modify the mappings legitimately.

I would revise this whole thought for clarity and meaning.

Author: Yeah no this is just bad wording in general. What i meant is that for a web shell to execute code the file ext must match a handler mapping that routes the request to an executable
handler. That requirement is what is immutable. However, an attacker can modify the handler mappings through 'web.config' which changes the essential operations and falls under proc C.

### ASP vs. ASP.NET Execution

Classic ASP files (`.asp`) are interpreted directly by `asp.dll` inside
`w3wp.exe`. No additional artifacts are produced on disk.

ASP.NET files (`.aspx`) undergo a compilation step. On first request, the
ASP.NET engine compiles the source into a .NET assembly (DLL) stored in
`C:\Windows\Microsoft.NET\Framework64\<version>\Temporary ASP.NET Files\`.

> On .NET Framework 4.8 (widely deployed), this compilation spawns `csc.exe` as

Reviewer: ```suggestion
On .NET Framework 4.8, this compilation spawns `csc.exe` as
```

Suggest removing `(widely deployed)` as it adds nothing

a child process of `w3wp.exe`. On ASP.NET Core with the Roslyn in-process

> compiler, compilation may occur without a child process.

Reviewer: You are bringing up `ASP.NET Core`, but you listed this as specifically out-of-scope in the beginning of the document. Make up your mind! Pick one and stay consistent. I would recommend leaving it out-of-scope as you are correct in the scope above that `ASP.NET Core` uses a different execution model altogether

> The compilation produces several artifacts in Temporary ASP.NET Files: a

Reviewer: ```suggestion
The compilation produces several artifacts in `Temporary ASP.NET Files`: a
```

Adding backticks around folder name for readability

`.compiled` metadata file, the compiled assembly (`.dll`), intermediate C#
source files (`.cs`), and compiler I/O files. The `.compiled` file's name

> preserves the original source extension — e.g., `shell.aspx.cdcab7d2.compiled`

Reviewer 2: ```suggestion
preserves the original source extension - e.g., `shell.aspx.cdcab7d2.compiled`
```

Another em dash

or `readme.txt.cdcab7d2.compiled` if a `.txt` file has been configured for
ASP.NET execution via handler manipulation (see Procedure C).

### The Web Root and Virtual Directories

Every IIS site maps to a physical directory (default: `C:\inetpub\wwwroot\`).
Virtual directories can extend the web-accessible area to other paths. For a
web shell to be reachable via HTTP, the shell file must reside within a
web-accessible path.

### web.config and Handler Manipulation

IIS configuration can be modified at the directory level through `web.config`
files. A `web.config` placed in any web-accessible directory overrides IIS
settings for that directory and its subdirectories.

An attacker who can write a `web.config` can manipulate handler mappings in two
ways. First, they can add a custom handler mapping that routes a normally static
extension (`.jpg`, `.txt`, `.png`) through the ASP.NET engine, then place a
web shell with that extension. Second, they can define an inline `IHttpHandler`

> within the `web.config` itself — no separate script file needed.

Reviewer 2: ```suggestion
within the `web.config` itself - no separate script file needed.
```

Em dash

The custom handler mapping variant requires two configuration elements: a
`<handlers>` entry routing the extension to `PageHandlerFactory`, and a
`<buildProviders>` entry registering `PageBuildProvider` for the extension.
Without both, ASP.NET fails with a compilation error. A practical complication
is that `system.webServer/handlers` is locked by default
(`overrideModeDefault="Deny"` in `applicationHost.config`), so subdirectory
`web.config` files cannot add handler entries to this section without an
administrator unlocking it. The `system.web/httpHandlers` section is not locked
by default, making it the more commonly exploitable path. The

> `<buildProviders>` element can only be defined at the IIS Application level —

Reviewer: This claim about the "more commonly exploitable path" intrigued me so I did some digging.

According to [Microsoft documentation](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.administration.applicationpool.managedpipelinemode?view=iis-dotnet), the `ApplicationPool.ManagedPipelineMode` property defaults to `Integrated Mode`

According to [this random post](https://weblog.west-wind.com/posts/2007/oct/10/httpmodule-and-httphandler-sections-in-iis-7-webconfig-files):

> you've probably seen the IIS 7 Exception that lets you know that in Integrated mode in IIS 7 you are not supposed to have an httpModules or httpHandlers section in <system.web>

This is also confirmed in [official Microsoft docs](https://learn.microsoft.com/en-us/iis/configuration/system.webserver/validation).

I think this section needs more clarification about `Integrated Mode` and `Classic Mode`, and what is possible with each. The TRR doesn't go into the distinction between the two modes and perhaps it should to cover the behavior differences.

Author: Yup, this is 100% a blind spot for me. Thanks! Just had to get a deeper crash course in the two modes. I think I read about them in my trr fever dream but it was one of many things lost to the brain void. I'll address it.

> `<buildProviders>` element can only be defined at the IIS Application level —

Reviewer 2: ```suggestion
`<buildProviders>` element can only be defined at the IIS Application level -
```

Em dash

ASP.NET returns a configuration error if it appears below the application root.
This means the custom handler variant requires either that the target directory
is already an IIS Application, the attacker can modify the site root's

> `web.config`, or the attacker can modify `machine.config` (requires SYSTEM

Reviewer 2: ```suggestion
`web.config`, or the attacker can modify `machine.config` (requires `SYSTEM`
```

I would consider `SYSTEM` to be a pronoun or keyword or whatever.

access).

When a `web.config` is placed in a subdirectory, IIS dynamically reloads that
directory's configuration without a server restart.

A third handler manipulation path exists: deploying a pre-compiled
`IHttpHandler` DLL to the application's `bin/` directory with a matching
`web.config` handler registration. Because the assembly is already compiled,
ASP.NET does not invoke `csc.exe` and does not produce the `App_Web_*.dll` or

> `.compiled` artifacts in Temporary ASP.NET Files that characterize the other

Reviewer 2: ```suggestion
`.compiled` artifacts in `Temporary ASP.NET Files` that characterize the other
```

Directory name.

variants. However, the CLR shadow-copies the DLL from `bin/` into
`Temporary ASP.NET Files\<app>\<hash>\assembly\dl3\` at load time, which is a
file creation event. This variant requires both a `web.config` handler entry
and write access to the `bin/` directory.

### IIS Application Initialization

IIS Application Initialization can be configured through `web.config` to send
internal warmup requests to specified pages on app pool start, recycle, or
reboot. If an attacker designates a web shell as a preload page, IIS will

> automatically trigger it on every lifecycle event — transforming the shell from

Reviewer 2: ```suggestion
automatically trigger it on every lifecycle event - transforming the shell from
```

Em dash

a passive backdoor into one that auto-executes without an attacker-initiated
request.

## Procedures

| ID | Title | Tactic |
|----|-------|--------|
| TRR0000.WIN.A | Web Shell with OS Command Execution | Persistence |
| TRR0000.WIN.B | Web Shell with In-Process Execution | Persistence |
| TRR0000.WIN.C | Web Shell via web.config Manipulation | Persistence |

> ### Procedure A: Web Shell with OS Command Execution

Reviewer: Procedure A currently describes `.aspx` and `.asp` as the shell vectors and includes "Compile ASPX" as a conditional sub-operation for `.aspx`. However, the "Technical Background" section establishes that SSI `#exec cmd=` directives also result in a child process spawn from `w3wp.exe`.... the same essential operation that defines this procedure.

While the current wording might suffice for a detection strategy, it falls short if a Red Team was to use this TRR for implementation of the procedue. The three execution paths that lead to that shared `w3wp.exe` spawning are meaningfully different in the following ways:

- ASP.NET (`.aspx`, `.ashx`, `.asmx`)
  - Compiled by the ASP.NET engine
  - Produces `csc.exe` spawn and artifacts in `Temporary ASP.NET Files`
  - Requires the ASP.NET role service
- Classic ASP (`.asp`)
  - Interpreted by `asp.dll`
  - No compilation artifacts
  - Requires the ASP role service (not installed by default, separate from ASP.NET)
- SSI (`.shtml`, `.stm`, `.shtm`)
  - Parsed by `ssinc.dll`
  - No scripting engine involvement
  - Attacker control limited to the command string in the directive
  - Requires the ServerSideIncludes role service

Consider documenting these as named variants under Procedure A (similar to how Procedure C handles its three variants), each with its own:
 
- Prerequisites
- Handler Path
- Upstream Artifacts
 
Try to do this while keeping them grouped under the shared essential operations. This would give a Red Team operator enough information to implement each path independently without breaking apart the procedural structure that the DDM is built on.

And yes.... I am aware DDM stands for "Detection Data Model", but the application of a DDM/TRR goes beyond just detection engineering.... a point I bring up with @vanvleeta often

Author: Yeah, I completely agree with this. If it's not obvious by now the SSI stuff was something I'd been struggling with fitting into the trr but I see it now.

A malicious script file (`.aspx` or `.asp`) is placed in a web-accessible

> directory — either as a new file or injected into an existing one. The delivery

Reviewer 2: ```suggestion
directory - either as a new file or injected into an existing one. The delivery
```

Em dash

method is tangential; the essential prerequisite is that malicious code exists
on disk in a path IIS is configured to serve.

When the attacker requests the shell's URL, the request enters the IIS
pipeline, is routed to `w3wp.exe`, matched to the appropriate handler, and
executed. The web shell then calls a process-creation API (e.g.,
`System.Diagnostics.Process.Start()`) to spawn an external program — typically

> `cmd.exe` or `powershell.exe` — with attacker-supplied arguments. The child

Reviewer 2: ```suggestion
`System.Diagnostics.Process.Start()`) to spawn an external program - typically
`cmd.exe` or `powershell.exe` - with attacker-supplied arguments. The child
```

Em dashes

process executes the command and the output is returned in the HTTP response.

The Process Spawn operation distinguishes this procedure from Procedure B.

#### Detection Data Model

![DDM - Web Shell with OS Command Execution](ddms/trr0000_win_a.png)

File prerequisites feed into the shared pipeline (Route Request → Match Handler

> → Execute Code). The distinguishing operation is Process Spawn from `w3wp.exe`.

Reviewer 2: ```suggestion
File prerequisites feed into the shared pipeline (Route Request -> Match Handler
-> Execute Code). The distinguishing operation is Process Spawn from `w3wp.exe`.
```

I'll defer to @vanvleeta, but this is more unicode. I imagine we'll see more of this with AI generated TRRs and it might just be something I have to live with.

Author: I fixed it, and added rules into the workflow and prompts to prevent unicode in the future. I know to look for unicode now. This trr was done with a human heavy assistance prompt with claude. While it was written by claude I've gone through it personally since the first research note. The unicode stuff just wasn't something I was aware would be an issue. So yeah it probably will come up again, but, it can avoided with the right rules in the workflow.

Compile ASPX is a sub-operation of Execute Code, relevant only for `.aspx`
shells.

> ### Procedure B: Web Shell with In-Process Execution

Reviewer: You write that Procedure B shells operate "exclusively through .NET framework APIs or COM objects within w3wp.exe."

COM objects are a broad category. Some COM objects are used for in-process operations. Examples include:

- `ADODB.Connection` for database access
- `Scripting.FileSystemObject` for file operations

Those fit Procedure B cleanly but one that comes to my Red Team brain is `WScript.Shell`. That is also a COM object and it's one of the most common ways web shells execute OS commands (via `.Run()` or `.Exec()`).

That spawns a child process from w3wp.exe, which is the essential operation that defines Procedure A, not B.

This may be nitpicky, but can we tighten the COM object language to exclude process-spawning objects like `WScript.Shell`..... or explicitly state that COM objects used for process creation fall under Procedure A regardless of the invocation method?

This procedure shares the same prerequisites and pipeline as Procedure A
through Execute Code. The difference is that the web shell never spawns a child

> process — it operates exclusively through .NET framework APIs or COM objects

Reviewer 2: ```suggestion
process - it operates exclusively through .NET framework APIs or COM objects
```

Em dash

within `w3wp.exe` (e.g., `System.IO.File.ReadAllText()`,
`System.Net.WebClient`, `System.Data.SqlClient`).

The specific API calls are attacker-controlled and tangential. The side effects
of those calls may produce telemetry depending on the action taken.

SSI directives such as `#include` and `#echo` also operate without spawning a
child process, but they do not map to this procedure. These directives are
processed by `ssinc.dll`'s built-in logic rather than through a scripting
engine executing attacker-authored code; they never pass through the Execute
Code operation in the pipeline.

#### Detection Data Model

![DDM - Web Shell with In-Process Execution](ddms/trr0000_win_b.png)

Identical to Procedure A through Execute Code. Diverges at the final step:
Call .NET API instead of Process Spawn. The Call .NET API operation has no
direct telemetry — only its side effects may be observable.

> ### Procedure C: Web Shell via web.config Manipulation

Reviewer: Your say all three Procedure C variants have the same general prerequisite:

> all three variants require writing a web.config to a web-accessible directory.

That's technically true, but it glosses over the fact that the custom handler mapping variant has a stricter write location requirement than the other two. The inline handler variant and pre-compiled DLL variant can work from a subdirectory `web.config` (assuming `system.webServer`/`handlers` is unlocked or the attacker uses `system.web`/`httpHandlers` (see the Integrated mode topic I raised earlier)). The custom handler mapping variant requires the application root.

If a Red Team operator had to implement this procedure, it would help if they could differentiate the write location prerequisite across the three variants. Specifically, calling out that the custom handler mapping variant requires application-root-level `web.config` access, not just any web-accessible directory.

This procedure introduces Write Config as an essential operation not present in
Procedures A or B. Write Config changes the essential operation chain in two
ways: it removes the constraint that the file extension must match a
default handler mapping, and in the inline `IHttpHandler` variant, it
eliminates the separate script file prerequisite entirely.

Three variants exist. In the custom handler mapping variant, the attacker
writes a `web.config` with both a handler mapping and build provider
registration, then places a shell with a normally static extension. The build
provider must be defined at the IIS Application level. In the inline handler
variant, the attacker defines an `IHttpHandler` directly in

> the `web.config` — no separate file needed. In the

Reviewer 2: ```suggestion
the `web.config` - no separate file needed. In the
```

Em dash

pre-compiled DLL variant, the attacker places a compiled
`IHttpHandler` assembly in the application's `bin/` directory with a matching
`web.config` handler registration; this bypasses `csc.exe` and standard
compilation artifacts, though the CLR shadow copy into
`Temporary ASP.NET Files\assembly\dl3\` still produces a file creation event.

All three variants require writing a `web.config` to a web-accessible
directory. The downstream pipeline operates the same as Procedures A and B;
post-execution behavior depends on the shell's code.

Application Initialization can further enhance persistence by configuring IIS
to auto-trigger the shell on app pool lifecycle events (see Technical

> Background).

Reviewer 2: ```suggestion
to auto-trigger the shell on app pool lifecycle events (see [Technical
Background]).
```

Use new global link below for reader's convenience.

When ASP.NET compiles a non-standard extension (custom handler variant), the
`.compiled` metadata file preserves the original extension in its filename
(e.g., `readme.txt.cdcab7d2.compiled`).

#### Detection Data Model

![DDM - Web Shell via web.config Manipulation](ddms/trr0000_win_c.png)

Write Config feeds into Match Handler, reflecting its modification of handler
matching. In the inline variant, Write Config is the sole prerequisite. In the
custom handler mapping and pre-compiled DLL variants, it is accompanied by a
file operation. The downstream pipeline and post-execution branches remain the
same as Procedures A and B.

## Available Emulation Tests

| ID             | Link                |
|----------------|---------------------|
| TRR0000.WIN.A  | [Atomic Test T1505.003-1], [Atomic Test T1505.003-2] |
| TRR0000.WIN.B  |                     |
| TRR0000.WIN.C  |                     |

> ## References

Reviewer: I would love to see [this microsoft docs page](https://learn.microsoft.com/en-us/previous-versions/aspnet/ms366723(v=vs.100)) which  goes into detail on how each of the ASP.NET files are compiled and what their artifacts are on disk.

- [IIS Architecture Overview - Microsoft Learn]
- [Handler Mappings in IIS - Microsoft Learn]
- [Application Pools in IIS - Microsoft Learn]
- [ASP.NET Compilation Overview - Microsoft Learn]
- [Virtual Directories in IIS - Microsoft Learn]
- [web.config Reference - Microsoft Learn]
- [IIS Application Initialization - Microsoft Learn]
- [Server-Side Includes in IIS - Microsoft Learn]
- [IHttpHandler Interface - Microsoft Learn]
- [Detect and Prevent Web Shell Malware - NSA/CISA]
- [Web Shell Attacks Continue to Rise - Microsoft Security Blog]
- [Ghost in the Shell - Microsoft Security Blog]
- [Web Shell Detection - Elastic Security]
- [Mo' Shells Mo' Problems: Deep Panda Web Shells - CrowdStrike]
- [T1505.003 - MITRE ATT&CK]

> 

Reviewer 2: ```suggestion

[Technical Background]: #technical-background
```

So you can link back to the technical background.

attacker-controlled parameters - it does not interpret or execute
attacker-authored code. [Procedure B] requires arbitrary code execution through
a scripting engine within `w3wp.exe`.

For a web shell to execute, its file extension must match a handler mapping that
routes the request to an executable handler rather than serving it as static
content. The requirement that a request match a handler mapping is an essential
operation in the pipeline - it cannot be skipped. However, the mappings
themselves are configurable. When an attacker modifies handler mappings through
`web.config` to route additional extensions through executable handlers, the
essential operations change - this is covered in [Procedure C].

### ASP vs. ASP.NET Execution

Classic ASP files (`.asp`) are interpreted directly by `asp.dll` inside
`w3wp.exe`. No additional artifacts are produced on disk.

ASP.NET files (`.aspx`) undergo a compilation step. On first request, the
ASP.NET engine compiles the source into a .NET assembly (DLL) stored in
`C:\Windows\Microsoft.NET\Framework64\<version>\Temporary ASP.NET Files\`. On
.NET Framework 4.8, this compilation spawns `csc.exe` as a child process of

> `w3wp.exe`.

Reviewer 3: Is this specific to that version - 4.8?

Author: Good call, at this point it's been months, but I think the version talk came from the rabbit hole of reading MS docs and confusion. Did a bit of digging through my notes to see how that got in there. Anyway, no it is not version specific. Which means I actually need to make two small changes and remove this v4.8 talk. Thanks!

pipeline mode determines which configuration section governs handler mappings:
Integrated mode reads from `system.webServer/handlers`, while Classic mode reads
from `system.web/httpHandlers` and ignores `system.webServer/handlers`. In
Integrated mode, entries in `system.web/httpHandlers` are not processed and
generate a migration error unless `validateIntegratedModeConfiguration` is set
to `false` in the application's `web.config`.

The custom handler mapping variant requires two configuration elements: a
`<handlers>` entry routing the extension to `PageHandlerFactory`, and a
`<buildProviders>` entry registering `PageBuildProvider` for the extension.
Without both, ASP.NET fails with a compilation error.

In Integrated mode, the attacker must target `system.webServer/handlers`, which
is locked by default (`overrideModeDefault="Deny"` in `applicationHost.config`).
Subdirectory `web.config` files cannot add handler entries to this section
unless an administrator has unlocked it or the attacker has sufficient access to
modify `applicationHost.config` directly. In Classic mode, the attacker targets
`system.web/httpHandlers`, which is not locked by default - but the application

> pool must be running in Classic mode rather than the default Integrated mode.

Reviewer 3: Just curious - how prevalent is that? Do a lot of people/orgs use Classic mode? What would be the reasons to do so? 

Update: answered my own question. Looks like it's very rare these days, as Integrated has been the default since 2008. Classic would only be used for a (very old) legacy scenario.

Author: Yeah probably not a worry anymore except at that one community college that never got state funding for upgrading tech.

ASP.NET does not invoke `csc.exe` and does not produce the `App_Web_*.dll` or
`.compiled` artifacts in `Temporary ASP.NET Files` that characterize the other
variants. However, the CLR shadow-copies the DLL from `bin/` into `Temporary
ASP.NET Files\<app>\<hash>\assembly\dl3\` at load time, which is a file creation
event. This variant requires both a `web.config` handler entry and write access
to the `bin/` directory.

### IIS Application Initialization

IIS Application Initialization can be configured through `web.config` to send
internal warmup requests to specified pages on app pool start, recycle, or
reboot. If an attacker designates a web shell as a preload page, IIS will
automatically trigger it on every lifecycle event - transforming the shell from
a passive backdoor into one that auto-executes without an attacker-initiated
request.

## Procedures

| ID | Title | Tactic |
|----|-------|--------|
| TRR0000.WIN.A | Web Shell with OS Command Execution | Persistence |
| TRR0000.WIN.B | Web Shell with In-Process Execution | Persistence |
| TRR0000.WIN.C | Web Shell via `web.config` Manipulation | Persistence |

> ### Procedure A: Web Shell with OS Command Execution

Reviewer 3: This section for Procedure A is great, but I feel like maybe it's missing information about how to get command execution, for example, with a malicious .aspx file. Maybe a small, simple example for each variant could be helpful?

Reviewer 3: I see that the section for Procedure B mentions "`WScript.Shell.Run()` or `WScript.Shell.Exec()`"...example of that kind of stuff could be good.

Reviewer 3: Maybe just small code excerpts. An entire file is probably not necessary. Obviously feel free to disagree

Author: Somehow this comment led me down a rabbit hole about SSI... again. I have to make more changes.

In the SSI variant, a file with a `.shtml`, `.stm`, or `.shtm` extension
containing an `#exec cmd=` directive is parsed by `ssinc.dll`. The Server Side
Includes role service must be installed. No scripting engine is involved -
`ssinc.dll` processes the directive directly and spawns a child process to
execute the command string. Attacker control is limited to the command string
within the directive.

All three variants converge at the same essential operation: the child process
spawned by `w3wp.exe` executes the command and output is returned in the HTTP
response.

#### Detection Data Model

![DDM - Web Shell with OS Command Execution](ddms/trr0000_win_a.png)

File prerequisites feed into the shared pipeline (Route Request, Match Handler,
Execute Code). Execute Code represents the handler processing the request within
`w3wp.exe` - through a scripting engine for the ASP.NET and Classic ASP
variants, and through `ssinc.dll` directive processing for the SSI variant. The
distinguishing operation is Process Spawn from `w3wp.exe`, shared across all
three variants. Compile ASPX is a sub-operation of Execute Code, relevant only
to the ASP.NET variant.

> ### Procedure B: Web Shell with In-Process Execution

Reviewer 3: Same as with Procedure A - maybe an example of what it would look like could be good?
